### ISSUE TITLE
Enabling `no-unused-variable` can trigger `no-unsafe-any` warnings

### ISSUE THREAD

**Issue by blair** [palantir/tslint#3863]
### Bug Report

- __TSLint version__: 5.9.1, although it also fails using a newer build master built from https://github.com/palantir/tslint/commit/8e13390e0a255d27695154af65e62002050e7adb
- __TypeScript version__: 2.8.3
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
import { BigNumber } from "bignumber.js";
const a = new BigNumber(10);
const b = a.times(2);
console.log(b.toString());
```

with `tslint.json` configuration:

```json
{
    "defaultSeverity": "error",
    "extends": [
        "tslint:recommended"
    ],
    "jsRules": {},
    "rules": {
        "no-console": false,
        "no-unused-variable": true,
        "no-unsafe-any": true
    },
    "rulesDirectory": []
}
```

#### Actual behavior
It generated these warnings:

```
ERROR: 2:15  no-unsafe-any  Unsafe use of expression of type 'any'.
ERROR: 3:11  no-unsafe-any  Unsafe use of expression of type 'any'.
ERROR: 4:13  no-unsafe-any  Unsafe use of expression of type 'any'.
```

#### Expected behavior

Expected behavior is no warnings. If one disables the `no-unused-variable` warning then the `no-unsafe-any` warnings disappear, which seems odd. Additionally, I don't see why the code as is generates warnings.

The code snippet is the smallest code I've been able to produce to generate the error.

Here's a shell script that will set up a new module and set up everything in it

```sh
#!/bin/sh

dir="tslint-bug.$$"
rm -fr "${dir}"
mkdir "${dir}"
cd "${dir}"

yarn init -y
yarn add -E -D typescript@2.8.3
yarn add -E -D tslint@5.9.1
yarn add -E bignumber.js@7.0.1

cat >tsconfig.json <<EOF
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es2017",
    "lib": ["dom", "es2017"],
    "strict": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "experimentalDecorators": true,
    "outDir": "build",
    "sourceMap": true,
    "declaration": true
  },
  "include": [
    "src/**/*.ts"
  ],
  "exclude": [
    "node_modules"
  ]
}
EOF

cat >tslint.json <<EOF
{
    "defaultSeverity": "error",
    "extends": [
        "tslint:recommended"
    ],
    "jsRules": {},
    "rules": {
        "no-console": false,
        "no-unused-variable": true,
        "no-unsafe-any": true
    },
    "rulesDirectory": []
}
EOF

mkdir src
cat >src/AvailableBalances.ts <<EOF
import { BigNumber } from "bignumber.js";
const a = new BigNumber(10);
const b = a.times(2);
console.log(b.toString());
EOF

./node_modules/.bin/tslint -c tslint.json -p . -t stylish
```

**Comment by maxx07** [comment]
I also have problem with that issue. I recognized that if: 
- the typescript definition file inside the **bignumber.js** module is renamed from `bignumber.d.ts` to `index.d.ts`
- the `types` property from `package.json` from that module is removed

the "no-unsafe-any" warning does not occur anymore.

If that is done the the default location (module root path and `index.d.ts` as name) for the typescript definition file is used, it seems there is no problem anymore.
But for sure according to [Typescript Handbook](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html) is must be possible using the `types/typings` property in the `package.json` file with any path to the typescript definition file.
So if using the script from the [comment before](https://github.com/palantir/tslint/issues/3863#issue-318734483) to reproduce the issue just add the following lines to the end of the script (or just call it in the created `tslint-bug.*` directory) to create the case described here that leads to a tslint run without a "no-unsafe-any" warning:
```
cp ./node_modules/bignumber.js/bignumber.d.ts ./node_modules/bignumber.js/index.d.ts
sed -i.bak '/bignumber\.d\.ts/d' ./node_modules/bignumber.js/package.json
```
and call it again:
```
./node_modules/.bin/tslint -c tslint.json -p . -t stylish
```

**Comment by giladgray** [comment]
hmm this seems like a problem with the bignumber.js package's typings, causing tsc to think `BigNumber` is `any` type. this is an external issue.

**Comment by jaufgang** [comment]
@giladgray, the fact that invalid typings in a third party library is what triggered this anomalous behaviour doesn't mean that there is not a tslint bug here, and I would suggest that it was a mistake to close the issue without investigating further. 

It doesn't explain why enabling the `no-unused-variable` rule should suddenly cause `no-unsafe-any` warnings to pop up that were not there when only the `no-unsafe-any` rule is enabled without `no-unused-variable`.  

If tsc thinks `BigNumber` is `any`, why doesn' tslint give a `no-unsafe-any` warning when `no-unused-variable` is turned off?  

I have experienced similar behaviour in my project.  Without `no-unused-variable` turned on, I do not get any tslint errors or warnings.  When i turn it on `no-unused-variable`, I get the following tslint errors:

- **1006** `Unsafe use of expression of type 'any'.` 
- **36**  `This assertion is unnecessary since it does not change the type of the expression.` 
- **193**  actual warnings about unused variables

I have looked at many of the unsafe-any and unnecessary type assertion errors, and they seem to be false positives, and none of them  make use of the BigNumber library.

**Comment by marhub** [comment]
i have similar error, when i disable no-unused-variable, i get "Promises must be handled appropriately", 
if i enable no-unused-variable, error is gone.
